src/algorithm/nest_cfgs.rs

```diff
@@ -71,6 +74,92 @@ pub trait CfgView<T> {
     fn successors(&self, node: T) -> Self::Iterator<'_>;
     /// Returns an iterator over the predecessors of the specified basic block.
     fn predecessors(&self, node: T) -> Self::Iterator<'_>;
+
+    /// Given an entry edge and exit edge defining a SESE region, mutates the
+    /// Hugr such that all nodes between these edges are placed in a nested CFG.
+    /// Returns the newly-constructed block (containing a nested CFG), or an error
+    /// if the two edges do not constitute a SESE region.
+    fn nest_sese_region(&mut self, entry_edge: (T, T), exit_edge: (T, T)) -> Result<T, String>;
+}
+
+/// Transforms a CFG to nested form.
+pub fn transform_cfg_to_nested<T: Copy + Eq + Hash + std::fmt::Debug>(
+    view: &mut impl CfgNodeMap<T>,
+) -> Result<(), String> {
+    let edge_classes = EdgeClassifier::get_edge_classes(view);
+    let mut rem_edges: HashMap<usize, HashSet<(T, T)>> = HashMap::new();
+    for (e, cls) in edge_classes.iter() {
+        rem_edges.entry(*cls).or_default().insert(*e);
+    }
+
+    // Traverse. Any traversal will encounter edges in SESE-respecting order.
+    fn traverse<T: Copy + Eq + Hash + std::fmt::Debug>(
```

Comment:
This has to be recursive because the calls with the `stop_at` parameter are too complex to use the internal `node_stack`. I tried using [fix_fn](https://crates.io/crates/fix_fn) to make a recursive closure (which would avoid passing `edge_classes` and `rem_edges`) but that doesn't allow declaring type parameters or using them from the containing method :-(